### PR TITLE
[c10d] Profiler support for nccl p2p collectives

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56840 Clang format ProcessGroupNCCL.cpp
* **#56427 [c10d] Profiler support for nccl p2p collectives**

This PR enables support for nccl send/recv profiling similar to how we have it for MPI and Gloo.

The process to do so is similar to the NCCL collectives where we create the `recordingFunction` in `initWork` and then add a callback that runs the profiler end callbacks. Tests are added similar to send/recv tests with gloo/MPI.

We also test with both autograd profiler and torch.profiler.

Profiler output:

```
----------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                    Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
----------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                cudaFree         8.95%     134.888ms         8.95%     134.888ms     134.888ms       0.000us         0.00%       0.000us       0.000us             1  
            cudaStreamCreateWithPriority         0.07%       1.060ms         0.07%       1.060ms      16.562us       0.000us         0.00%       0.000us       0.000us            64  
                cudaEventCreateWithFlags         0.00%      22.000us         0.00%      22.000us       3.667us       0.000us         0.00%       0.000us       0.000us             6  
                   cudaDeviceGetPCIBusId         0.00%      21.000us         0.00%      21.000us       1.909us       0.000us         0.00%       0.000us       0.000us            11  
                           cudaHostAlloc         0.05%     710.000us         0.05%     710.000us     355.000us       0.000us         0.00%       0.000us       0.000us             2  
                  cudaDeviceGetAttribute         0.00%       1.000us         0.00%       1.000us       0.250us       0.000us         0.00%       0.000us       0.000us             4  
                              cudaMalloc         2.41%      36.381ms         2.41%      36.381ms       6.064ms       0.000us         0.00%       0.000us       0.000us             6  
                              cudaMemset        87.30%        1.316s        87.30%        1.316s     219.251ms       0.000us         0.00%       0.000us       0.000us             6  
                         Memset (Device)         0.00%       0.000us         0.00%       0.000us       0.000us      70.000us         1.44%      70.000us      11.667us             6  
                      cudaGetDeviceCount         0.00%       1.000us         0.00%       1.000us       0.167us       0.000us         0.00%       0.000us       0.000us             6  
                 cudaDeviceCanAccessPeer         0.00%       5.000us         0.00%       5.000us       2.500us       0.000us         0.00%       0.000us       0.000us             2  
                     cudaIpcGetMemHandle         0.00%      45.000us         0.00%      45.000us      22.500us       0.000us         0.00%       0.000us       0.000us             2  
                    cudaIpcOpenMemHandle         0.03%     427.000us         0.03%     427.000us     213.500us       0.000us         0.00%       0.000us       0.000us             2  
                              cudaMemcpy         0.00%      74.000us         0.00%      74.000us      14.800us       0.000us         0.00%       0.000us       0.000us             5  
        Memcpy HtoD (Pageable -> Device)         0.00%       0.000us         0.00%       0.000us       0.000us       6.000us         0.12%       6.000us       1.000us             6  
                         cudaEventRecord         0.00%      11.000us         0.00%      11.000us       1.375us       0.000us         0.00%       0.000us       0.000us             8  
                     cudaStreamWaitEvent         0.00%      10.000us         0.00%      10.000us       1.429us       0.000us         0.00%       0.000us       0.000us             7  
                               nccl:send         0.08%       1.272ms         0.54%       8.150ms       8.150ms      19.000us         0.39%      19.000us      19.000us             1  
                        cudaLaunchKernel         0.46%       6.886ms         0.46%       6.886ms       3.443ms       0.000us         0.00%       0.000us       0.000us             2  
    ncclSendRecvKernel_copy_i8(ncclColl)         0.00%       0.000us         0.00%       0.000us       0.000us       4.772ms        98.41%       4.772ms       2.386ms             2  
                      record_param_comms         0.00%      46.000us         0.00%      51.000us      25.500us       0.000us         0.00%       0.000us       0.000us             2  
                        cudaEventDestroy         0.00%       3.000us         0.00%       3.000us       0.750us       0.000us         0.00%       0.000us       0.000us             4  
                             aten::empty         0.03%     389.000us         0.03%     389.000us      32.417us       0.000us         0.00%       0.000us       0.000us            12  
                             aten::fill_         0.01%     148.000us         0.01%     148.000us      49.333us       0.000us         0.00%       0.000us       0.000us             3  
                                aten::to         0.04%     574.000us         0.34%       5.157ms     644.625us       0.000us         0.00%       2.000us       0.250us             8  
                     aten::empty_strided         0.01%     205.000us         0.01%     205.000us      41.000us       0.000us         0.00%       0.000us       0.000us             5  
                             aten::copy_         0.02%     228.000us         0.29%       4.378ms     875.600us       2.000us         0.04%       2.000us       0.400us             5  
                         cudaMemcpyAsync         0.27%       4.137ms         0.27%       4.137ms       2.068ms       0.000us         0.00%       0.000us       0.000us             2  
                   cudaStreamSynchronize         0.00%      13.000us         0.00%      13.000us       6.500us       0.000us         0.00%       0.000us       0.000us             2  
                               nccl:recv         0.01%     116.000us         0.01%     138.000us     138.000us       4.753ms        98.02%       4.753ms       4.753ms             1  
        Memcpy DtoH (Device -> Pageable)         0.00%       0.000us         0.00%       0.000us       0.000us       1.000us         0.02%       1.000us       1.000us             1  
                     aten::promote_types         0.00%      23.000us         0.00%      23.000us      23.000us       0.000us         0.00%       0.000us       0.000us             1  
                          aten::allclose         0.01%     134.000us         0.29%       4.443ms       4.443ms       0.000us         0.00%       0.000us       0.000us             1  
                           aten::isclose         0.03%     499.000us         0.26%       3.934ms       3.934ms       0.000us         0.00%       0.000us       0.000us             1  
                                aten::eq         0.02%     304.000us         0.03%     446.000us     111.500us       0.000us         0.00%       0.000us       0.000us             4  
                                aten::ne         0.04%     532.000us         0.08%       1.272ms     212.000us       0.000us         0.00%       0.000us       0.000us             6  
                          aten::__iand__         0.01%      89.000us         0.02%     287.000us     143.500us       0.000us         0.00%       0.000us       0.000us             2  
                      aten::bitwise_and_         0.01%      91.000us         0.01%     198.000us      99.000us       0.000us         0.00%       0.000us       0.000us             2  
                       aten::bitwise_and         0.01%     107.000us         0.01%     107.000us      53.500us       0.000us         0.00%       0.000us       0.000us             2  
                           aten::__ior__         0.01%      80.000us         0.02%     287.000us     143.500us       0.000us         0.00%       0.000us       0.000us             2  
                       aten::bitwise_or_         0.01%     111.000us         0.01%     207.000us     103.500us       0.000us         0.00%       0.000us       0.000us             2  
                        aten::bitwise_or         0.01%      96.000us         0.01%      96.000us      48.000us       0.000us         0.00%       0.000us       0.000us             2  
                               aten::mul         0.02%     241.000us         0.03%     395.000us     197.500us       0.000us         0.00%       0.000us       0.000us             2  
                               aten::abs         0.03%     419.000us         0.04%     645.000us     107.500us       0.000us         0.00%       0.000us       0.000us             6  
                               aten::add         0.01%     141.000us         0.02%     310.000us     310.000us       0.000us         0.00%       0.000us       0.000us             1  
                               aten::sub         0.00%      74.000us         0.00%      74.000us      74.000us       0.000us         0.00%       0.000us       0.000us             1  
                          aten::isfinite         0.01%     160.000us         0.07%     985.000us     985.000us       0.000us         0.00%       0.000us       0.000us             1  
                              aten::item         0.00%      66.000us         0.01%     123.000us      61.500us       0.000us         0.00%       0.000us       0.000us             2  
               aten::_local_scalar_dense         0.00%      57.000us         0.00%      57.000us      28.500us       0.000us         0.00%       0.000us       0.000us             2  
                                aten::le         0.01%     169.000us         0.02%     238.000us     119.000us       0.000us         0.00%       0.000us       0.000us             2  
                               aten::all         0.01%     226.000us         0.02%     310.000us     310.000us       0.000us         0.00%       0.000us       0.000us             1  
                        aten::as_strided         0.00%      36.000us         0.00%      36.000us      36.000us       0.000us         0.00%       0.000us       0.000us             1  
----------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 1.507s
Self CUDA time total: 4.849ms
```

Differential Revision: [D27866600](https://our.internmc.facebook.com/intern/diff/D27866600/)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1159,12 +1160,20 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::pointToPoint(
   syncStreams(devices, ncclEvents_[key], ncclStreams_[key]);
 
   // Work itself will create the CUDA events on all GPUs of tensors
-  auto work = initWork(devices, rank_, opType);
+  bool can_profile = tensors.size() == 1;
+  auto work = initWork(
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:1164]: nit and this is prior to this PR: looks like `initWork` didn't follow clang-format convention? (the last arg line should be split to two). Could you please run a clang-format in a followup PR? 

```
c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL> ProcessGroupNCCL::initWork(
    std::vector<at::Device> devices,
    int rank,
    OpType opType,
    const char* profilingTitle, const c10::optional<std::vector<at::Tensor>>& inputs) {
  return c10::make_intrusive<ProcessGroupNCCL::WorkNCCL>(devices, rank, opType, profilingTitle, inputs);
}
```

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -1159,12 +1160,20 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::pointToPoint(
   syncStreams(devices, ncclEvents_[key], ncclStreams_[key]);
 
   // Work itself will create the CUDA events on all GPUs of tensors
-  auto work = initWork(devices, rank_, opType);
+  bool can_profile = tensors.size() == 1;
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:1163]: shall we add some comments to explain why we have this restriction?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -1208,13 +1217,26 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::pointToPoint(
...
-  if (opType == OpType::RECV) {
+
+  // Future only needs to be created and marked completed with outputs for
+  // recv(), but still create future for use cases such as profiling even for
+  // send().
+  {
     at::cuda::CUDAMultiStreamGuard streamGuard(ncclStreams_[key]);
     work->future_ = c10::make_intrusive<at::cuda::CUDAFuture>(
         c10::ListType::create(c10::TensorType::get()));
     work->future_->markCompleted(at::IValue(*work->outputs_));
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:1228]: Will this change the behavior? If this is `OpType::SEND`, what's in `*work->outputs_`?

Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1228]: For `send()` it will be the input tensors. I think previously, we didn't have this future at all which is sort of strange because ideally for nccl, `work.get_future()` should return a future regardless of the type of work, but before this it would be an "empty" future. 

With this PR, if we do:
```
tensor = torch.tensor(...)
fut = ncclSend(tensor).get_future()
fut.wait() # will just return tensor unmodified
```

Actually, we should probably add some tests for `get_future()` for p2p nccl collectives cc @agolynski

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1163]: This is because when https://github.com/pytorch/pytorch/pull/46471/files that originally enabled the profiling was added, we used NCCLFuture instead of CUDAFuture which only supported single device per process. 

Now that we use CUDAFuture it might be possible this restriction can be removed. Will add comment and file issue for that.

Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1163]: https://github.com/pytorch/pytorch/issues/56836

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (mrshenli) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1236]: Is this behavior documented anywhere?

**Review by mrshenli — APPROVED** [review]
LGTM

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1236]: We don't document `get_future` anywhere and only mention it for gradient compression currently: https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html?highlight=get_future. I think after future/work merge and `get_future()` API having support for all backends, we should add documentation for this. cc @agolynski

**Comment by rohan-varma** [comment]
Failure is unrelated - 

```
May 05 02:28:14 Expected equality of these values:
May 05 02:28:14   at_tensor[0][2].item().to<float>()
May 05 02:28:14     Which is: 1
May 05 02:28:14   345
May 05 02:28:14 [  FAILED  ] CUDAPytorchToCaffe2.MutualResizes (1 ms)
May 05 02:28:14 [----------] 3 tests from CUDAPytorchToCaffe2 (3 ms total)
May 05 02:28:14 
May 05 02:28:14 [----------] Global test environment tear-down
May 05 02:28:14 [==========] 6 tests from 2 test cases ran. (5196 ms total)
May 05 02:28:14 [  PASSED  ] 5 tests.
May 05 02:28:14 [  FAILED  ] 1 test, listed below:
```

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:1232]: Filed https://github.com/pytorch/pytorch/issues/57646 for this
